### DISCUSSION TITLE
Start Containers only if needed

### DISCUSSION THREAD

**Discussion by Zomono** [Trendyol/jest-testcontainers#38] [Ideas]
Hi,
jest-testcontainers always starts all containers specified in the global jest-testcontainer-config.js file. Most of my tests do not need any container while the other ones only need one or two containers.
If you are running all test together this doesn't matter, but the usual workflow is to run only a single test or a single test suit. This workflow is slowed down by waiting for all containers to finish start up.

So my idea is to tell jest-testcontainers on a test-file basis which containers are reuqired. If no container is required, no one will be started.
This could be done be calling a function with container name as argument eg: useContainer("redis").

However, this library is already awesome. Thanks!

**Comment by Yengas** [comment]
Hey @Zomono, 

Currently this library bootstraps everything before your tests are actually run. If I understood you correctly; you would like to specify which containers to run inside your test files. This would be nice to have but it could complicate the flow of the library because we would need to invent some way of passing information from tests file to the initial bootstrap/setup process. We could of course do this somehow, but it could complicate things unnecessarily.

However as an alternative (without complicating library itself), you could make use of the fact that config file is actually a JavaScript file. For example; inside the `jest-testcontainers-config.js` you can look at the environment variables, program arguments etc. to figure out which tests will be run and expose your configuration accordingly.

IMHO implementing some generic way to handle this in this library would be much harder for us. E.g. the system should support TS/JS, it should support configs with docker-compose / plain configurations etc. etc.

Let me know what you think! Also thanks for your kind words.

**Reply by Zomono** [reply]
My idea would be to do all the setup the same way as it currently is, except for the "start container and wait for it"-part. This last step could be lazily done for each container when any test asks for the container and be skipped, if it is alreay running. Therefore, each test has to declare which container it needs in some way.
Actually, the biggest problem for this feature is the sandboxing applied for each test by jest. I don't know how to share the state (ip, port) of an already running container. 
However, it seems that testcontainers for java has already implemented this feature, where you can call "withReuse" on the GenericContainer class. Maybe node-testcontainers follows this example.

**Reply by Yengas** [reply]
@Zomono as you mention currently this library starts the containers in an eager manner. The configuration file is the source of truth for which containers to run. In global setup of jest; this configuration file is read, containers are started and the random ports are written to some file. Test Environment feature of jest is used to read the port / host information from the files and set environment variables.

For this reason; implementing something like **withReuse** could be harder and change this flow too much. Because in that case the code itself becomes the source of truth. As you mentioned it seems to me like it is more appropiate for [testcontainers-node](https://github.com/testcontainers/testcontainers-node) to implement this.

One approach I see that could be an optional feature to this library is; you can create/find a library that would scan all of the test files to be run and look for some specificly structured comments. Then you can use the information in these comments to return a different configuration in your jest-testcontainers Javascript configuration file. E.g.

```
// file: *.test.js
// @trendyol/jest-testcontainers { needs: ['mysql', 'redis'] }
``` 

```
// file: jest-testcontainers-config.js
...
const testRequirements = analyzer.collectTestRequirements();

module.exports = {
  ...(testRequirements.shouldRun('mysql') ? { ...someConfig } : {})
}
```
